substrate/client/network/src/litep2p/shim/notification/mod.rs

```diff
@@ -0,0 +1,370 @@
+			.field("protocol", &self.protocol)
+			.field("handle", &self.handle)
+			.finish()
+	}
+}
+
+impl NotificationProtocol {
+	/// Create new [`NotificationProtocol`].
+	pub fn new(
+		protocol: ProtocolName,
+		handle: NotificationHandle,
+		peerset: Peerset,
+		metrics: NotificationMetrics,
+	) -> Self {
+		Self {
+			protocol,
+			handle,
+			peerset,
+			metrics,
+			pending_cancels: HashSet::new(),
+			pending_validations: FuturesUnordered::new(),
+		}
+	}
+
+	/// Handle `Peerset` command.
+	async fn on_peerset_command(&mut self, command: PeersetNotificationCommand) {
+		match command {
+			PeersetNotificationCommand::OpenSubstream { peers } => {
+				log::debug!(target: LOG_TARGET, "{}: open substreams to {peers:?}", self.protocol);
+
+				let _ = self.handle.open_substream_batch(peers.into_iter().map(From::from)).await;
+			},
+			PeersetNotificationCommand::CloseSubstream { peers } => {
+				log::debug!(target: LOG_TARGET, "{}: close substreams to {peers:?}", self.protocol);
+
+				self.handle.close_substream_batch(peers.into_iter().map(From::from)).await;
+			},
+		}
+	}
+}
+
+#[async_trait::async_trait]
+impl NotificationService for NotificationProtocol {
+	async fn open_substream(&mut self, _peer: PeerId) -> Result<(), ()> {
+		unimplemented!();
+	}
+
+	async fn close_substream(&mut self, _peer: PeerId) -> Result<(), ()> {
+		unimplemented!();
+	}
+
+	fn send_sync_notification(&mut self, peer: &PeerId, notification: Vec<u8>) {
+		let size = notification.len();
+
+		if let Ok(_) = self.handle.send_sync_notification(peer.into(), notification) {
+			self.metrics.register_notification_sent(&self.protocol, size);
+		}
+	}
+
+	async fn send_async_notification(
+		&mut self,
+		peer: &PeerId,
+		notification: Vec<u8>,
+	) -> Result<(), Error> {
+		let size = notification.len();
+
+		match self.handle.send_async_notification(peer.into(), notification).await {
+			Ok(_) => {
+				self.metrics.register_notification_sent(&self.protocol, size);
+				Ok(())
+			},
+			Err(_) => Err(Error::ChannelClosed),
+		}
+	}
+
+	/// Set handshake for the notification protocol replacing the old handshake.
+	async fn set_handshake(&mut self, handshake: Vec<u8>) -> Result<(), ()> {
+		self.handle.set_handshake(handshake);
+
+		Ok(())
+	}
+
+	/// Non-blocking variant of `set_handshake()` that attempts to update the handshake
+	/// and returns an error if the channel is blocked.
+	///
+	/// Technically the function can return an error if the channel to `Notifications` is closed
+	/// but that doesn't happen under normal operation.
+	fn try_set_handshake(&mut self, handshake: Vec<u8>) -> Result<(), ()> {
+		self.handle.set_handshake(handshake);
+
+		Ok(())
+	}
+
+	/// Make a copy of the object so it can be shared between protocol components
+	/// who wish to have access to the same underlying notification protocol.
+	fn clone(&mut self) -> Result<Box<dyn NotificationService>, ()> {
+		unimplemented!("clonable `NotificationService` not supported by `litep2p`");
```

Comment:
This function is used in tests because there are a few GRANDPA tests that clone the `NotificationService`. Those tests need to rewritten before this method can be removed